### PR TITLE
Disable external link for default and pre-screener programs in North Star

### PR DESCRIPTION
**Pull request by EmiliaPaz**
### Description

External program link is used by:
- Default and pre-screener programs in pre-NS (optional)
- External programs in NS (required)

Therefore, this PR disables the 'external link' field for default and pre-screener programs in North Star and makes it a required value on for external programs.

Disabled fields are readonly, thus if the test tries to add a value to them the locator times out. `addProgram()` in `browser-test/src/support/admin_programs.ts` adds default values to all fields. This PR is disabling the `external link` field for default programs and pre-screeners in NS. Thus, to avoid the tests failing we:
  - Add `addPreScreenerNS()` and `addExternalProgram()` that only add values for fields that are required
  - In `addProgram()`, we check for value existence before adding it to optional or disabled fields. Ideally we should do the same for external link, but that required updating lots of tests so leaving as an issue for now (see comment in the code)
  - Re-ordered field filling to mimic the order in the form. Specifically, we want to first fill the program type before the fields which are disabled based on program type

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label (see [docs](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-appropriate-labels) for more info): < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [ ] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [ ] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).
- [ ] Ensured PII wasn't added to any new logs, unless it was guarded by `isDevOrStaging`
- [ ] Noted in the PR description which, if any, code in this PR was generated by AI.

#### User visible changes

- [ ] Followed steps to [internationalize](https://github.com/civiform/civiform/wiki/Internationalization-%28i18n%29#application-strings) any new strings
  - [ ] Added context strings to new [messages](https://github.com/civiform/civiform/blob/main/server/conf/i18n/messages)
  - [ ] Didn't use a message in applicant facing code that isn't translated yet (unless behind a flag)
- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Made equivalent changes in Thymeleaf for applicant-facing features and changes (or created an issue in the [Northstar epic](https://github.com/orgs/civiform/projects/1/views/94) to track that it's missing from Thymeleaf)
- [x] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [x] Manually tested at 200% size
- [x] Manually evaluated tab order
- [ ] Manually tested with a screen reader if the feature is applicant-facing. See [screen reader testing](https://github.com/civiform/civiform/wiki/Testing#screen-reader-testing)

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/helper/feature-flags.conf`
- [x] Conditioned new functionality on a [FeatureFlag](https://github.com/civiform/civiform/wiki/Feature-Flags)
- [x] Wrote browser tests with the feature flag off and on, etc.

### Instructions for manual testing

If instructions are needed for manual testing by reviewers, include them here.

### Issue(s) this completes

Part of #10183

### CODE REVIEW

**Comment by EmiliaPaz** [comment]
* **#10599** <a href="https://app.graphite.dev/github/pr/civiform/civiform/10599?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/civiform/civiform/10599?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: server/app/assets/javascripts/admin_programs.ts [hunk 1]
```diff
@@ -103,16 +103,31 @@ class AdminPrograms {
...
+    const isNorthstarEnabled = externalLink.dataset.northstarEnabled === 'true'
+    const disableExternalLink =
+      (programType === ProgramType.DEFAULT ||
+        programType === ProgramType.COMMON_INTAKE_FORM) &&
+      isNorthstarEnabled
+    this.updateTextFieldElementDisabledState(
+      /* fieldElement= */ externalLink,
+      /* shouldDisable= */ disableExternalLink,
+    )
+    this.hideRequiredIndicators(
```
Comment (rockycodes) [inline server/app/assets/javascripts/admin_programs.ts:119]: suggestion (blocking): this is currently showing the required indicator for default and common intake programs when north star is disabled. this could be in it's own if check that just checks for `!ProgramType.EXTERNAL` since we never want the required indicator to show for other program types.

File: server/app/assets/javascripts/admin_programs.ts [hunk 2]
```diff
@@ -211,15 +230,28 @@ class AdminPrograms {
...
-    })
+    const labelElement = document.querySelector(fieldSelector)
+    if (!labelElement) {
+      return
+    }
+
+    let requiredSpan = labelElement.querySelector(
+      'span.text-red-600.font-semibold',
+    )
+    if (!requiredSpan) {
```
Comment (rockycodes) [inline server/app/assets/javascripts/admin_programs.ts:241]: praise/suggestion: great idea adding in the required spans when they're missing. i have two suggestions:
1. move this into it's own method for adding required spans that is not nested i this method since it's a separate concern. or change the name of this method to something more general like `handleRequiredIndicators`.
2. this is currently resulting in a funky state for application steps 
image

you could remove the "optional" label from step 1, or (my preference) make this change in a separate PR since it's a broader concern than just for external URLS.

File: browser-test/src/admin/admin_program_creation.test.ts [hunk 3]
```diff
@@ -1052,6 +1052,10 @@ test.describe('program creation', () => {
       await adminPrograms.expectFormFieldDisabled(FormField.PROGRAM_ELIGIBILITY)
       await adminPrograms.expectFormFieldDisabled(FormField.PROGRAM_CATEGORIES)
       await adminPrograms.expectFormFieldDisabled(FormField.APPLICATION_STEPS)
+      // Program external link is always enabled on pre North Star
+      await adminPrograms.expectFormFieldEnabled(
+        FormField.PROGRAM_EXTERNAL_LINK,
```
Comment (rockycodes) [inline browser-test/src/admin/admin_program_creation.test.ts:1057]: praise: thank you for relying on element checks instead of adding new screenshots!

**Review by rockycodes — CHANGES_REQUESTED** [review]
this looks great! i had some nitpicky changes related to test organization and a couple of suggestions for little bugs/edge cases i caught.

File: browser-test/src/support/admin_programs.ts [hunk 4]
```diff
@@ -240,40 +307,29 @@ export class AdminPrograms {
...
+    await this.page.fill('#program-name-input', slugify(programName))
+    await this.page.fill('#program-description-textarea', adminDescription)
 
     // Program type selector varies with the EXTERNAL_PROGRAM_CARDS feature.
     // When enabled, form has program type options. Otherwise, form has a
     // pre-screener checkbox.
+    // IMPORTANT: Select the program type first since some of the next fields
+    // are disabled based on the program type.
     const externalProgramsFeatureEnabled =
       await this.getProgramTypeOption(programType).isVisible()
```
[the remark this answers is not in the document]
Comment (EmiliaPaz) [reply browser-test/src/support/admin_programs.ts]: Great idea. Added a testId to the program type options field (used when external programs feature is enabled)

File: server/app/views/admin/programs/ProgramFormBuilder.java [hunk 5]
```diff
@@ -289,7 +291,13 @@ private FormTag buildProgramForm(
             .setId("program-external-link-input")
             .setFieldName("externalLink")
             .setLabelText("Link to program website")
+            .setRequired(isExternalProgram)
```
[the remark this answers is not in the document]
Comment (EmiliaPaz) [reply server/app/views/admin/programs/ProgramFormBuilder.java:295]: Hmm external programs were implemented under the assumption that North Star would be rolled out, thus we didn't add support for external programs not in NS.

I think it's fine to just leave with `isExternalProgram`, since external programs should always have an external link.

However, this highlighted a different bug. If admin added external programs on pre-NS they will appear to the user. Card will have wrong controls. We should not include external programs in pre-NS. Will tackle in a separate PR.
(I was hoping NS would be rolled out by the time we roll out external programs, however it sounds better to not be depending on the launches timelines)

image

[the remark this answers is not in the document]
Comment (melanie-exygy) [reply server/app/views/admin/programs/ProgramFormBuilder.java:295]: I think we could put the responsibility on the admins to only turn on the external programs flag in conjunction with the NS flag.  The description of the flag does say "On North Star applicant UI':  
image

**Review by melanie-exygy — APPROVED** [review]

**Review by melanie-exygy — APPROVED** [review]
